src/checker.rs

```diff
@@ -30,66 +30,70 @@ impl Checker {
             let res = client.get(uri).await;
+
             match res {
-                Ok(r) => {
-                    let u = &site.url[..];
-
-                    if self.allow_redirections && r.status().is_redirection() {
-                            println!("SUCCESS: {} for {}", r.status(), u);
-                            for (k, _) in &self.errors {
-                                if k.0 == &site.url[..] {
-                                    println!("unsetting error {}, {}", k.1, u);
-                                    INT_GAUGE_VECT.with_label_values(&[k.1.as_str(), u, self.prometheus_rule_scope.as_str()]).set(0);
-                                }
-                            }
-                            self.errors.retain(|key, _| {
-                                !(key.0 == &site.url[..])
-                            });
-                            INT_GAUGE_VECT.with_label_values(&["connection error", &site.url[..], self.prometheus_rule_scope.as_str()]).set(0);
+                Ok(resp) => {
+                    if self.allow_redirections && resp.status().is_redirection() {
+                            self.handle_success(&resp, url);
                     }
                     else{
-
-                        match r.status().as_str() {
+                        match resp.status().as_str() {
                             "200" | "301" | "308" => {
-                                println!("SUCCESS: {} for {}", r.status(), u);
-                                for (k, _) in &self.errors {
-                                    if k.0 == &site.url[..] {
-                                        println!("unsetting error {}, {}", k.1, u);
-                                        INT_GAUGE_VECT.with_label_values(&[k.1.as_str(), u, self.prometheus_rule_scope.as_str()]).set(0);
-                                    }
-                                }
-                                self.errors.retain(|key, _| {
-                                    !(key.0 == &site.url[..])
-                                });
-                                INT_GAUGE_VECT.with_label_values(&["connection error", &site.url[..], self.prometheus_rule_scope.as_str()]).set(0);
+                                self.handle_success(&resp, url);
                             },
-                            s => {
-                                println!("FAILURE: {} for {}", s, u);
-                                self.errors.insert((u.to_string(), s.to_string()), true);
-                                INT_GAUGE_VECT.with_label_values(&[s, u, self.prometheus_rule_scope.as_str()]).set(1);
+                            code => {
+                                println!("FAILURE: {} for {}", code, url);
+                                self.errors.insert((url.to_string(), code.to_string()), true);
+                                INT_GAUGE_VECT.with_label_values(&[code, url, self.prometheus_rule_scope.as_str()]).set(1);
```

Comment:
Bonus pints if you extract this plus Lines 87-89 to handle_failure method :)

Reply:
Done